core/src/test/java/io/substrait/type/proto/WindowFunctionRoundtripTest.java

```diff
@@ -0,0 +1,73 @@
+import io.substrait.expression.ImmutableExpression;
+import io.substrait.expression.ImmutableWindowBound;
+import io.substrait.extension.ExtensionCollector;
+import io.substrait.proto.Rel;
+import io.substrait.relation.ConsistentPartitionWindow;
+import io.substrait.relation.ImmutableConsistentPartitionWindow;
+import io.substrait.relation.ImmutableVirtualTableScan;
+import io.substrait.relation.ProtoRelConverter;
+import io.substrait.relation.RelProtoConverter;
+import io.substrait.relation.VirtualTableScan;
+import io.substrait.type.TypeCreator;
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.util.Arrays;
+import java.util.Collections;
+import org.junit.jupiter.api.Test;
+
+public class WindowFunctionRoundtripTest extends TestBase {
+
+  private void assertWindowFunctionRoundtrip(Expression.AggregationInvocation invocation) {
+    Expression.DecimalLiteral expression = ExpressionCreator.decimal(false, BigDecimal.TEN, 10, 2);
+    Expression.StructLiteral literal =
+        ImmutableExpression.StructLiteral.builder().from(expression).build();
+    ImmutableVirtualTableScan input = VirtualTableScan.builder().addRows(literal).build();
+    ExtensionCollector functionCollector = new ExtensionCollector();
+    RelProtoConverter to = new RelProtoConverter(functionCollector);
+    ProtoRelConverter from = new ProtoRelConverter(functionCollector, defaultExtensionCollection);
+
+    ImmutableConsistentPartitionWindow aggRel =
+        ImmutableConsistentPartitionWindow.builder()
+            .input(input)
+            .windowFunctions(
+                Arrays.asList(
+                    ConsistentPartitionWindow.WindowRelFunctionInvocation.builder()
+                        .declaration(defaultExtensionCollection.windowFunctions().get(0))
+                        .arguments(Collections.emptyList())
+                        .options(Collections.emptyMap())
+                        .outputType(TypeCreator.of(true).I64)
+                        .aggregationPhase(Expression.AggregationPhase.INITIAL_TO_RESULT)
+                        .invocation(invocation)
+                        .lowerBound(ImmutableWindowBound.Unbounded.UNBOUNDED)
+                        .upperBound(ImmutableWindowBound.Following.CURRENT_ROW)
+                        .boundsType(Expression.WindowBoundsType.RANGE)
+                        .build()))
+            .partitionExpressions(Arrays.asList(expression))
+            .sorts(
+                Arrays.asList(
+                    Expression.SortField.builder()
+                        .expr(expression)
+                        .from(b.sortField(expression, Expression.SortDirection.ASC_NULLS_FIRST))
+                        .build()))
+            .build();
+
+    Rel protoWindowRel = to.toProto(aggRel);
+    assertEquals(protoWindowRel, to.toProto(from.from(protoWindowRel)));
+    assertEquals(
+        protoWindowRel.getWindow().getWindowFunctions(0).getInvocation(), invocation.toProto());
```

Comment:
As written, this test doesn't catch the missing bounds conversion. I would recommend doing something like:

```java
    # note: whenever we're mixing proto and pojo code I find it useful to use fully qualified paths for clarity
    io.substrait.proto.Rel protoRel = to.toProto(aggRel);
    io.substrait.relation.Rel aggRel2 = from.from(protoRel);
    assertEquals(aggRel, aggRel2);
```

We should compare the example we build directly, in this case `aggRel` with the result of converting to and from proto. In this case the `toProto` conversion is lossy, but because we aren't checking against our built example we miss this.

Reply:
Thanks! It does catch the error and used it to verify the fix.